db/views/reporting_overdue_patients_v01.sql

```diff
@@ -0,0 +1,194 @@
+    LEFT JOIN appointments ON appointments.patient_id = rps.patient_id
+        AND appointments.device_created_at < rps.month_date
+    WHERE rps.status <> 'dead'
+	AND rps.month_date > NOW() - INTERVAL '24 months'
+    ORDER BY
+        rps.patient_id,
+        rps.month_date,
+        appointments.device_created_at DESC
+),
+
+patients_with_appointments_and_visits AS (
+    SELECT
+        DISTINCT ON ( patients_with_appointments.patient_id, patients_with_appointments.month_date)
+        patients_with_appointments.*,
+        visit_id,
+        visited_at_after_appointment
+    FROM patients_with_appointments
+    LEFT JOIN lateral (
+        SELECT id AS visit_id, patient_id, recorded_at AS visited_at_after_appointment
+        FROM blood_sugars
+        WHERE deleted_at IS NULL
+          AND patient_id = patients_with_appointments.patient_id
+          AND patients_with_appointments.previous_appointment_date < blood_sugars.recorded_at
+          AND blood_sugars.recorded_at < patients_with_appointments.month_date + INTERVAL '1 month' + INTERVAL '15 days'
+        UNION
+        (
+            SELECT id AS visit_id, patient_id, recorded_at AS visited_at_after_appointment
+            FROM blood_pressures
+            WHERE deleted_at IS NULL
+              AND patient_id = patients_with_appointments.patient_id
+              AND patients_with_appointments.previous_appointment_date < blood_pressures.recorded_at
+              AND blood_pressures.recorded_at < patients_with_appointments.month_date + INTERVAL '1 month' + INTERVAL '15 days'
+        )
+        UNION
+        (
+            SELECT id AS visit_id, patient_id, device_created_at AS visited_at_after_appointment
+            FROM appointments AS patients_with_appointments_visit
+            WHERE deleted_at IS NULL
+              AND patient_id = patients_with_appointments.patient_id
+              AND patients_with_appointments.previous_appointment_date < patients_with_appointments_visit.device_created_at
+              AND patients_with_appointments_visit.device_created_at < patients_with_appointments.month_date + INTERVAL '1 month' + INTERVAL '15 days'
+        )
+        UNION
+        (
+            SELECT id AS visit_id, patient_id, device_created_at AS visited_at_after_appointment
+            FROM prescription_drugs
+            WHERE deleted_at IS NULL
+              AND patient_id = patients_with_appointments.patient_id
+              AND patients_with_appointments.previous_appointment_date < prescription_drugs.device_created_at
+              AND prescription_drugs.device_created_at < patients_with_appointments.month_date + INTERVAL '1 month' + INTERVAL '15 days'
+        )
+    ) AS visits ON patients_with_appointments.patient_id = visits.patient_id
+    ORDER BY
+        patients_with_appointments.patient_id,
+        patients_with_appointments.month_date,
+        patients_with_appointments.previous_appointment_date DESC,
+        visited_at_after_appointment
```

Comment:
[This sort](https://explain.depesz.com/s/3Kkh#l15) seems to take up a long time and 6.5G. The memory consumption is probably because it's a composite key of 4.

I think this can be worked around since we already have one row per patient per month with a `previous_appointment_date` in `patients_with_appointments`. Something like

```
LEFT JOIN lateral (
     SELECT DISTINCT ON (patients_with_appointments.month_date)
     id AS visit_id, patient_id, recorded_at AS visited_at_after_appointment
     FROM blood_sugars
     WHERE ...
     UNION ... (the 4 unions)
     ORDER BY visited_at_after_appointment
   ) AS visits ON patients_with_appointments.patient_id = visits.patient_id
```

Reply:
Accepted